### DISCUSSION TITLE
Clarifying the usage of offset & LogStabilizationMoles

### DISCUSSION THREAD

**Discussion by ghost** [shannon-lab/zapdos#126] [Q&A Software Development] [answered]
I have a few quick questions regarding the role of LogStabilizationMoles & offset.

My current understanding of the LogStabilizationMoles kernel and the offset term is: **the log-molar density form in Zapdos is designed to prevent zero and negative density values. Nonetheless, for the sake of stability, an offset _b_ is added to assist values approaching zero. This value _b_ is added to the variable at each element, but it has the biggest impact where the variable is close to zero.** Please correct me if any of this is incorrect.

Some questions following this:

1. Does the offset term strictly add some log-molar density value _-b_ to every species variable across the whole domain? If not, what does the value _b_ correspond to, physically speaking, or is "add" not the right word to describe its function?
2. If _b_ indeed performs some addition function, is it only applied once, and thus its contribution is constant & uniform, or is it re-applied each timestep, like a source term with a constant rate over the whole domain? 

Please correct me if I have an incorrect conception of this kernel, and thank you for your help!

**Comment by csdechant** [comment, marked as answer]
Your second question is on the right track. `LogStabilizationMoles` acts as a source term in the form of _exp(-(b+u))_ (please keep in mind that _u_ here is the log form the the density, such that _n = exp(u)_, where _n_ is the density). Since this source term is depended on _exp(-u)_, it will be more impactful for small densities vs large densities. There is nothing physically to _b_, and it is used to offset _exp(-u)_ so that this artificial term does not full dominate the other source terms. _b_ depends on the problem set, but we found that a _b_ of 20 is usually sufficient. I should note that not all Zapdos simulation need this stabilization (e.g. @lindsayad and I believe @keniley1 need it for the 1D plasma-water interact simulations, but it was not required for the 1Torr argon cases I ran). I would advise trying running simulations without the stabilization, and if it seems that low densities are causing instabilities, then add `LogStabilizationMoles` to your simulations.

**Reply by ghost** [reply]
Thank you for confirming that I was on the right track! When I asked if _b_ had any physical meaning, I was mostly wondering if it had some associated density value: for example, if _b_ was 20, then it would add _exp(-b)  = exp(-20) = 2.061E-9_ mol/m^3 to each point where _u_ exists. Looking at how its formulated, exp(-(b+u)) it looks like it would exponentially increase the addition as b + u further drops below 0, add little to nothing if b + u = 0 or greater. A higher b value would then suggest it lets u decrease farther before acting as a significant source.

Is there some density _n_ that is just too low for Zapdos to handle, thus requiring the use of LogStabilizationMoles? My first guess is that in the plasma-liquid sims, some species' densities may go below this threshold at some point, but in the argon sims the density never goes below the threshold so you don't need the offset. Or is this picture too simplistic?

**Reply by csdechant** [reply]
For your first statement, I am not sure if assigning _b_  as some density value was the original vision for this stabilization method, but one could see it as such. Let _n_b_ be a density value, such that _n_b = exp(-b)_ if a positive _b_ value, then `LogStabilizationMoles` is define as _n_b*exp(-u)_  or _n_b / n_. So when _n ≥ n_b_, then the artificial source term is 1 or less and when _n < n_b_, then the artificial source term is greater then 1. So, in terms of _b_, a larger _b_ results in a smaller _n_b_. This means the artificial source term is 1 or less for a greater range of _n_ densities. (So yes, that was a round about way of agreeing with your statement of "A higher b value would then suggest it lets u decrease farther before acting as a significant source")

As for a "threshold", the species' densities are not the dominating factor. It is the energy density (i.e. _n_ε = ε * n_e_, where _n_ε_ is the energy density, _ε_ is the electron mean energy, and _n_e_ is the electron density). For the plasma-water simulations, the electron energy drops nearly to zero at the plasma-liquid interface, causing the instability. In the low pressure argon simulations, the electron energy is never that low. As for a limit of how low the electron energy can be before causing instabilities, I have study how the electron energy affects simulation results, but I have yet to set a "hard" limit to that.

**Reply by ghost** [reply]
Oh! The electron energy causes instabilities, not species' densities? That is very useful to know... Thank you for the golden nugget of knowledge and your help clarifying the kernel!
